perf/impl/https/v0.1/main.go

```diff
@@ -0,0 +1,252 @@
+package main
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/binary"
+	"encoding/json"
+	"encoding/pem"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"math/big"
+	"net"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+const (
+	BlockSize = 64 << 10
+)
+
+func handleRequest(w http.ResponseWriter, r *http.Request) {
+	// Read the big-endian bytesToSend value
+	var bytesToSend uint64
+	if err := binary.Read(r.Body, binary.BigEndian, &bytesToSend); err != nil {
+		http.Error(w, "failed to read uint64 value", http.StatusBadRequest)
+		return
+	}
+
+	// Read and discard the remaining bytes in the request
+	io.Copy(io.Discard, r.Body)
+
+	// Set content type and length headers
+	w.Header().Set("Content-Type", "application/octet-stream")
+	w.Header().Set("Content-Length", strconv.FormatUint(bytesToSend, 10))
+
+	// Write the status code explicitly
+	w.WriteHeader(http.StatusOK)
+
+	buf := make([]byte, BlockSize)
+
+	for bytesToSend > 0 {
+		toSend := buf
+		if bytesToSend < BlockSize {
+			toSend = buf[:bytesToSend]
+		}
+
+		n, err := w.Write(toSend)
+		if err != nil {
+			http.Error(w, "Failed write", http.StatusInternalServerError)
+			return
+		}
+		bytesToSend -= uint64(n)
+	}
+}
+
+var zeroSlice = make([]byte, BlockSize) // Pre-made zero-filled slice
+
+type customReader struct {
+	downloadBytes uint64
+	uploadBytes   uint64
+	position      uint64
+}
+
+func (c *customReader) Read(p []byte) (int, error) {
+	if c.position < 8 {
+		binary.BigEndian.PutUint64(p, c.downloadBytes)
+		c.position += 8
+		return 8, nil
```

Comment:
This isn't correct. You're assuming the input slice is always at least 8 bytes.

Use [`io.MultiReader`](https://cs.opensource.google/go/go/+/refs/tags/go1.20.4:src/io/multi.go;drc=946b4baaf6521d521928500b2b57429c149854e7;l=73) to combine a `bytes.NewReader(<buf-with-u64-download-bytes>)` with the and a new zeroReader.

The zeroReader could look something like:

```
type zeroReader struct {
    n int64
}

func (z *zeroReader) Read(p []byte) (int, error) {
    if z.n <= 0 {
        return 0, io.EOF
    }

    for i := range p {
        if z.n <= 0 {
            return i, nil
        }

        p[i] = 0
        z.n--
    }

    return len(p), nil
}
```

Reply:
Wasn't aware of `io.MultiReader`. That simplifies things. Thanks!

Fixed with [2296d8c](https://github.com/libp2p/test-plans/pull/184/commits/2296d8c1c8501bebe7a0f5f56379d6374245ebc6). Note that I am still using `copy` instead of zeroing every byte individually in the (maybe premature) hope that Go can optimize some of it. Let me know in case you prefer individual zeroing for each byte instead @MarcoPolo.